File: AlgoWorks/src/com/solid/algolearning/javacode/algorithms/sorting/BubbleSort.java
package com.solid.algolearning.javacode.algorithms.sorting;

import com.sun.org.apache.xpath.internal.objects.XNumber;

import java.util.Arrays;

public class BubbleSort {

//    Bubble sort implementation
    public static void main(String[] args) {
        int[] numbers = {34, 7, 65, 2, 8, 1, 2};

        System.out.println(Arrays.toString(bubbleSortIterative(numbers)));

        bubbleSortRecursive(numbers, 0, 0);
        System.out.println(Arrays.toString(numbers));
    }


    static int[] bubbleSortIterative(int[] numbers){
        for(int i = 0; i < numbers.length; i++){
            for (int j = i + 1; j < numbers.length; j++){
                if (numbers[i] > numbers[j]){
                    swap(numbers, i, j);
                }
            }
        }

        return numbers;
    }



    static void bubbleSortRecursive(int[] arr, int start, int end){
        if(start == 0) return;

        if (end < start){
            if(arr[start] > arr[end]){
                swap(arr, start, end);
            }
            bubbleSortRecursive(arr, start, end + 1);
        }else {
           bubbleSortRecursive(arr, start - 1, end);
        }
    }

    static void swap(int[] arr, int start, int end){
        int temp = arr[start];
        arr[start] = arr[end];
        arr[end] = temp;
    }
}

File: AlgoWorks/src/com/solid/algolearning/kotlincode/algorithms/sorting/BubbleSort.kt
package com.solid.algolearning.kotlincode.algorithms.sorting

//Bubble sort comes from the idea of bubbling up the largest value eto the right and is 0(n^2) time complexity
fun bubbleSort(arr:IntArray):IntArray{
    var swap = true
    while(swap){
        swap = false
        for(i in 0 until arr.size-1){
             if(arr[i] > arr[i+1]){
                val temp = arr[i]
                arr[i] = arr[i+1]
                arr[i + 1] = temp

                swap = true
            }
        }
    }
    return arr
}

fun main() {
    val list = bubbleSort(intArrayOf(2,15,1,8,4))
    for (k in list) print("$k ")
}